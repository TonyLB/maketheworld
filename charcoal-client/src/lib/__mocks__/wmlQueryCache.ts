export const wmlQueryFromCache = jest.fn()
