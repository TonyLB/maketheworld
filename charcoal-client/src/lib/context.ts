//
// context library holds utility functions for dealing with nested contexts and onChange functions.
//

import { SchemaTag } from "@tonylb/mtw-wml/dist/schema/baseClasses"
import { GenericTree, GenericTreeNode, GenericTreeNodeWithUndefined, GenericTreeWithUndefined } from "@tonylb/mtw-wml/dist/tree/baseClasses"
import { excludeUndefined } from "./lists"

export const nestOnChangeSubItem = <T extends SchemaTag>({ tree, index }: { tree: GenericTree<T>, index: number }) => (onChange: (newValue: GenericTree<T>) => void) => (newValue: GenericTreeNode<T>): void => {
    if (index < tree.length || index >= 0) {
        onChange([...tree.slice(0, index), newValue, ...tree.slice(index+ 1)])
    }
}

export const nestOnChangeChildren = <T extends SchemaTag>(node: GenericTreeNode<T>) => (onChange: (newValue: GenericTreeNode<T>) => void) => (newValue: GenericTree<T>): void => {
    onChange({ ...node, children: newValue })
}

//
// onChangeTreeReducer recursively builds an aggregate reducer to operate on a tree, having passed
// the sub-tree calculations through a functional transform at each level
//
type ChangeTreeReducer<T extends SchemaTag> = (previous: GenericTree<T> | undefined, newValue: GenericTree<T>) => GenericTree<T>
type ChangeTreeReducerInternal<T extends SchemaTag> = (previous: GenericTree<T> | undefined, newValue: GenericTreeWithUndefined<T>) => GenericTree<T>

export const nestTransformTreeReducer = <T extends SchemaTag, Supplement extends {}>(
    reducerTransform: (baseReducer: ChangeTreeReducer<T>, options: { parentData?: T }, addSupplemental: (supplement: Supplement) => void) => ChangeTreeReducerInternal<T>,
    baseReducer: ChangeTreeReducer<T> = (_, newValue) => (newValue),
    addSupplemental: (supplemental: Supplement) => void = () => {}
): ChangeTreeReducerInternal<T> => (previous, newValue): GenericTree<T> => {
    //
    // Recursively generate a new onChange function
    //
    const recursedOutput = reducerTransform(baseReducer, {}, addSupplemental)(previous, newValue.map((node, index) => {
        if (node) {
            const { data, children } = node
            return { data, children: nestTransformTreeReducer((baseReducer) => (reducerTransform(baseReducer, { parentData: data }, addSupplemental)), baseReducer, addSupplemental)(previous?.[index]?.children ?? undefined, children) }
        }
        else {
            return undefined
        }
    }))
    return recursedOutput.filter(excludeUndefined)
}