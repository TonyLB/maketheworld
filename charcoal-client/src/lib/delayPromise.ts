export const delayPromise = (delay: number) => {
    return new Promise<void>((resolve) => {
        setTimeout(() => { resolve() }, delay)
    })
}

export default delayPromise
