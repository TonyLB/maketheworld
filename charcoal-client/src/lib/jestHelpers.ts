export function mockFunction<T extends (...args: any[]) => any>(fn: T): jest.MockedFunction<T> {
    return fn as jest.MockedFunction<T>;
}

export function mockClass <T extends { new (...args: any): any }>(item: T): jest.MockedClass<typeof item> {
    return item as jest.MockedClass<typeof item>
}
