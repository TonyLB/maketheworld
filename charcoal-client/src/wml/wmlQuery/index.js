import wmlGrammar from '../wmlGrammar/wml.ohm-bundle.js'

import { wmlSelectorFactory } from './selector.js'
import { validatedSchema } from '../index.js'
import { normalize } from '../normalize.js'

const renderFromNode = (normalForm) => ({ tag, type, value = '', props = {}, contents = [] }) => {
    switch(type) {
        case 'tag':
            const flattenedProps = Object.entries(props)
                .reduce((previous, [key, { value }]) => ({ ...previous, [key]: value }), {})
            return {
                tag,
                ...flattenedProps,
                text: contents.map(renderFromNode(normalForm)).filter((item) => (typeof item !== 'object')).join(''),
                targetTag: normalForm[flattenedProps.to]?.tag === 'Action' ? 'Action' : 'Feature'
            }
        default:
            return value
    }
}

export class WMLQueryResult {
    constructor(wmlQuery, search) {
        this.wmlQuery = wmlQuery
        this.search = search
        this.refresh()
    }

    refresh() {
        const match = this.wmlQuery.matcher.match()
        if (match.succeeded()) {
            this._nodes = wmlSelectorFactory(match)(this.search)
        }
        else {
            this._nodes = []
        }
    }

    not(searchString) {
        const excludeResults = new WMLQueryResult(this.wmlQuery, searchString)
        const excludeStarts = excludeResults.nodes().map(({ start }) => (start))
        this._nodes = this._nodes.filter(({ start }) => (!excludeStarts.includes(start)))
        return this
    }

    replaceInputRange(startIdx, endIdx, value) {
        this.wmlQuery.replaceInputRange(startIdx, endIdx, value)
    }

    get source() {
        return this.wmlQuery.source
    }

    nodes() {
        return this._nodes || []
    }

    remove() {
        let offset = 0
        this._nodes.forEach(({ start, end }) => {
            this.replaceInputRange(start - offset, end - offset, '')
            offset += end - start
        })
        this.refresh()
        return this
    }

    children() {
        this._nodes = this._nodes.reduce((previous, { contents = [] }) => ([...previous, ...contents]), [])
        return this
    }

    prepend(sourceString) {
        if (this._nodes.length) {
            this.replaceInputRange(this._nodes[0].start, this._nodes[0].start, sourceString)
            this.refresh()
        }
        return this
    }

    prop(key, value) {
        if (value !== undefined) {
            this._nodes.forEach((node) => {
                if (node.props[key]) {
                    const { valueStart, valueEnd } = node.props[key]
                    if (valueEnd) {
                        this.replaceInputRange(valueStart, valueEnd, value)
                    }
                }
                else {
                    const insertAfter = Object.values(node.props || {})
                        .reduce((previous, { end }) => (Math.max(previous, end)), node.tagEnd)
                    this.replaceInputRange(insertAfter, insertAfter, ` ${key}="${value}"`)
                }
            })
            this.refresh()
            return this
        }
        else {
            if (this._nodes.length) {
                return this._nodes[0].props[key]?.value
            }
            return undefined
        }
    }

    removeProp(key) {
        this._nodes.forEach((node) => {
            if (node.props[key]) {
                const { start, end } = node.props[key]
                if (end) {
                    this.replaceInputRange(start-1, end, '')
                }
            }
        })
        this.refresh()
        return this
    }

    contents(value) {
        if (value !== undefined) {
            this._nodes.forEach((node) => {
                const { start, end } = (node.contents || []).reduce((previous, probeNode) => ({
                    start: Math.min(probeNode.start, previous.start),
                    end: Math.max(probeNode.end, previous.end)
                }), { start: node.end, end: 0 })
                if (end > 0) {
                    this.replaceInputRange(start, end, value)
                }
            })
            this.refresh()
            return this
        }
        else {
            if (this._nodes.length) {
                return this._nodes[0].contents || []
            }
            return []
        }
    }

    render(value) {
        if (value !== undefined) {
            const renderContents = value.map((item) => {
                if (typeof item === 'object') {
                    return `<Link key=(${item.key}) to=(${item.to})>${item.text}</Link>`
                }
                else {
                    return item
                }
            })
            let offset = 0
            this._nodes.forEach((node) => {
                //
                // Generate new contents, replacing current render elements with new
                //
                const nonRenderNodes = node.contents.filter(({ type, tag }) => (type === 'tag' && tag !== 'Link'))

                const reApplyNonRenderContents = nonRenderNodes.map(({ start, end }) => (this.source.substring(start, end)))
                const revisedContents = [
                    ...renderContents,
                    ...reApplyNonRenderContents
                ].join("\n")
                //
                // Calculate the entire span being filled with contents
                //
                const { start, end } = (node.contents || []).reduce((previous, probeNode) => ({
                    start: Math.min(probeNode.start, previous.start),
                    end: Math.max(probeNode.end, previous.end)
                }), { start: node.end, end: 0 })
                if (end > 0) {
                    this.replaceInputRange(start+offset, end+offset, revisedContents)
                    offset += revisedContents.length - (end - start)
                }
            })
            this.refresh()
            return this
        }
        else {
            if (this._nodes.length && ['Room', 'Feature'].includes(this._nodes[0].tag)) {
                return (this._nodes[0].contents || [])
                    .filter(({ tag, type }) => (
                        type === 'string' || tag === 'Link'
                    ))
                    .map(renderFromNode(this.wmlQuery.normalize()))
            }
            return []
        }
    }
}

export class WMLQuery {
    constructor(sourceString, options = {}) {
        const { onChange = () => {} } = options
        this.onChange = onChange
        this.matcher = wmlGrammar.matcher()
        this.matcher.setInput(sourceString)
    }

    get source() {
        return this.matcher.getInput()
    }
    setInput(str) {
        this.matcher.setInput(str)
        this.onChange({
            type: 'set',
            text: str,
            wmlQuery: this
        })
    }
    normalize() {
        const schema = validatedSchema(this.matcher.match())
        return normalize(schema)
    }
    replaceInputRange(startIdx, endIdx, str) {
        this.matcher.replaceInputRange(startIdx, endIdx, str)
        this.onChange({
            type: 'replace',
            startIdx,
            endIdx,
            text: str,
            wmlQuery: this
        })
    }

    search(searchString) {
        return new WMLQueryResult(this, searchString)
    }
}
