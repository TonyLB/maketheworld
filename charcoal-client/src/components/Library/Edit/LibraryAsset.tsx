/** @jsxImportSource @emotion/react */

//
// LibraryAsset is a context provider (with associated useLibraryAsset
// context subscriber hook) to create component nests that all operate in the
// context of having a chosen asset that a player is examining in the library.
//
// Arguments:
//
//   - AssetId: The internal key of the asset.
//
// Context provided:
//
//   - AssetId
//

import React, { useContext, ReactChild, ReactChildren, FunctionComponent, useMemo, useCallback, useState, useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'

import {
    getCurrentWML,
    getImportDefaults,
    getNormalized,
    getLoadedImages,
    setIntent,
    getProperties,
    updateNormal as updateNormalAction,
    getDraftWML
} from '../../../slices/personalAssets'
import { heartbeat } from '../../../slices/stateSeekingMachine/ssmHeartbeat'
import { NormalForm, NormalComponent, ComponentRenderItem, NormalExit, isNormalExit, isNormalComponent } from '@tonylb/mtw-wml/dist/normalize/baseClasses'
import { objectFilter } from '../../../lib/objects'
import { AssetClientImportDefaults } from '@tonylb/mtw-interfaces/dist/asset'
import { PersonalAssetsLoadedImage } from '../../../slices/personalAssets/baseClasses'
import { getConfiguration } from '../../../slices/configuration'
import { UpdateNormalPayload } from '../../../slices/personalAssets/reducers'

type LibraryAssetContextType = {
    assetKey: string;
    AssetId: string;
    currentWML: string;
    draftWML: string;
    normalForm: NormalForm;
    importDefaults: AssetClientImportDefaults["defaultsByKey"];
    updateNormal: (action: UpdateNormalPayload) => void;
    loadedImages: Record<string, PersonalAssetsLoadedImage>;
    properties: Record<string, { fileName: string }>;
    components: Record<string, AssetComponent>;
    rooms: Record<string, AssetComponent>;
    exits: Record<string, NormalExit>;
    features: Record<string, AssetComponent>;
    save: () => void;
}

const LibraryAssetContext = React.createContext<LibraryAssetContextType>({
    assetKey: '',
    AssetId: '',
    currentWML: '',
    draftWML: '',
    normalForm: {},
    importDefaults: {},
    updateNormal: () => {},
    properties: {},
    loadedImages: {},
    components: {},
    rooms: {},
    exits: {},
    features: {},
    save: () => {}
})

type LibraryAssetProps = {
    assetKey: string;
    children?: ReactChild | ReactChildren;
    character?: boolean;
}

export type AssetComponent = {
    tag: string;
    localName: ComponentRenderItem[];
    localRender: ComponentRenderItem[];
    defaultName: ComponentRenderItem[];
    defaultRender?: ComponentRenderItem[];
    name: ComponentRenderItem[];
    render: ComponentRenderItem[];
}

const assetComponents = ({ normalForm, importDefaults }: { normalForm: NormalForm, importDefaults: AssetClientImportDefaults["defaultsByKey"] }): Record<string, AssetComponent> => {
    const componentNormals = Object.values(normalForm).filter((item) => (isNormalComponent(item))) as NormalComponent[]

    const roomReturns = componentNormals
        .map((component) => {
            const localName = (component.appearances
                .filter(({ contextStack }) => (!contextStack.find(({ tag }) => (tag === 'If'))))
                .map(({ name = [] }) => name)
                .reduce((previous, name) => ([ ...previous, ...name ]), [])
                .map((item) => ((item.tag === 'String') ? item.value : ''))
                .join('')) || ''
            const countRenderAppearances = component.appearances
                .filter(({ contextStack }) => (!contextStack.find(({ tag }) => (tag === 'If'))))
                .length
            const localRender = component.appearances
                .filter(({ contextStack }) => (!contextStack.find(({ tag }) => (tag === 'If'))))
                .map(({ render = [] }) => render)
                .reduce((previous, render) => ([ ...previous, ...render ]), [])
            const defaultName = importDefaults[component.key]?.Name || []
            const defaultRender = importDefaults[component.key]?.Description || []
            return { [component.key]: {
                tag: component.tag,
                localName,
                localRender,
                defaultName,
                defaultRender,
                name: [ ...defaultName, ...localName ],
                render: [...(defaultRender || []), ...localRender]
            }}
        })
    
    return Object.assign({}, ...roomReturns)
}

export const LibraryAsset: FunctionComponent<LibraryAssetProps> = ({ assetKey, children, character }) => {

    const AssetId = useMemo<string>(() => (`${character ? 'CHARACTER' : 'ASSET'}#${assetKey}`), [character, assetKey])
    const currentWML = useSelector(getCurrentWML(AssetId))
    const draftWML = useSelector(getDraftWML(AssetId))
    const normalForm = useSelector(getNormalized(AssetId))
    const importDefaults = useSelector(getImportDefaults(AssetId))
    const loadedImages = useSelector(getLoadedImages(AssetId))
    const properties = useSelector(getProperties(AssetId))
    const dispatch = useDispatch()
    const updateNormal = useCallback((updateAction: UpdateNormalPayload) => {
        dispatch(updateNormalAction(AssetId)(updateAction))
        dispatch(setIntent({ key: AssetId, intent: ['NORMALDIRTY'] }))
        dispatch(heartbeat)
    }, [dispatch, AssetId])
    const components = useMemo<Record<string, AssetComponent>>(() => ( assetComponents({ normalForm, importDefaults }) ), [normalForm, importDefaults])
    const rooms = useMemo<Record<string, AssetComponent>>(() => ( objectFilter(components, ({ tag }) => (tag === 'Room')) ), [components])
    const exits = useMemo<Record<string, NormalExit>>(() => ( objectFilter(normalForm, isNormalExit) ), [components])
    const features = useMemo<Record<string, AssetComponent>>(() => ( objectFilter(components, ({ tag }) => (tag === 'Feature')) ), [components])
    const save = useCallback(() => {
        dispatch(setIntent({ key: AssetId, intent: ['NEEDSAVE'] }))
        dispatch(heartbeat)
    }, [dispatch, AssetId])

    return (
        <LibraryAssetContext.Provider value={{
            assetKey,
            AssetId,
            currentWML,
            draftWML,
            normalForm,
            importDefaults,
            updateNormal,
            properties,
            loadedImages,
            components,
            rooms,
            exits,
            features,
            save
        }}>
            {children}
        </LibraryAssetContext.Provider>
    )
}

export const useLibraryAsset = () => (useContext(LibraryAssetContext))

type ImageHeaderSyntheticURL = {
    loadId: string;
    fileURL: string;
}

export const useLibraryImageURL = (key: string): string => {    
    const { loadedImages, properties } = useLibraryAsset()
    const { AppBaseURL = '' } = useSelector(getConfiguration)
    const [syntheticURL, setSyntheticURL] = useState<ImageHeaderSyntheticURL | undefined>()

    const loadedImage = useMemo(() => (
        loadedImages[key]
    ), [loadedImages, key])

    useEffect(() => {
        if (loadedImage?.loadId !== syntheticURL?.loadId) {
            if (syntheticURL) {
                URL.revokeObjectURL(syntheticURL.fileURL)
            }
            setSyntheticURL({
                loadId: loadedImage.loadId,
                fileURL: URL.createObjectURL(loadedImage.file)
            })
        }
        return () => {
            if (syntheticURL) {
                URL.revokeObjectURL(syntheticURL.fileURL)
            }
        }
    }, [syntheticURL, loadedImage])

    const fileURL = useMemo(() => {
        const appBaseURL = process.env.NODE_ENV === 'development' ? `https://${AppBaseURL}` : ''
        return syntheticURL ? syntheticURL.fileURL : properties[key] ? `${appBaseURL}/images/${properties[key].fileName}.png` : ''
    }, [syntheticURL, properties, key])

    return fileURL

}

export default LibraryAsset
