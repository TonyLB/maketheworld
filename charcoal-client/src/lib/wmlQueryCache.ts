import { WMLQuery } from '@tonylb/mtw-wml/dist/wmlQuery'

let wmlQueryCacheArray: Record<string, WMLQuery> = {}

export const wmlQueryFromCache = ({ key, value = '' }: { key: string, value?: string }): WMLQuery => {
    if (wmlQueryCacheArray[key]) {
        if (value) {
            wmlQueryCacheArray[key].setInput(value)
        }
        return wmlQueryCacheArray[key]
    }
    const returnValue = new WMLQuery(value || '')
    wmlQueryCacheArray[key] = returnValue
    return returnValue
}
