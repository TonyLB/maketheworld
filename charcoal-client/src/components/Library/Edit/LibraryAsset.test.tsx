import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import LibraryAsset, { useLibraryAsset, AssetComponent } from './LibraryAsset'

jest.mock('../../../cacheDB')
jest.mock('../../../lib/wmlQueryCache')
import { wmlQueryFromCache } from '../../../lib/wmlQueryCache'
import { WMLQuery } from '@tonylb/mtw-wml/dist/wmlQuery'

const wmlQueryFromCacheMock = wmlQueryFromCache as jest.Mock

const mockStore = configureStore()
const currentWML = `
<Asset key=(Test) fileName="test">
    <Import from=(BASE)>
        <Use key=(basePower) type="Variable" as=(power) />
        <Use key=(DEF) type="Room" />
    </Import>
    <Feature key=(clockTower)>
        <Description>
            A worn stone clock tower.
        </Description>
    </Feature>
    <Room key=(ABC)>
        <Name>Vortex</Name>
        <Description>
            Vortex
            <Link to=(toggleOpen)>(toggle)</Link>
        </Description>
        <Exit from=(DEF)>vortex</Exit>
    </Room>
    <Condition if={open}>
        <Depend on=(open) />
        <Room key=(ABC)>
            <Exit to=(DEF)>welcome</Exit>
        </Room>
    </Condition>
    <Room key=(DEF)>
        <Name>Welcome</Name>
    </Room>
    <Variable key=(open) default={false} />
    <Action key=(toggleOpen) src={open = !open} />
</Asset>
`

const store = mockStore({
   personalAssets: {
       byId: {
           ['ASSET#Test']: {
               publicData: {
                   originalWML: currentWML,
                   currentWML,
                   importDefaults: {
                        DEF: {
                            Description: [{
                                tag: 'String',
                                value: 'A welcome area'
                            }],
                            Name: []
                        }
                   }
               },
               meta: {
                   currentState: 'FRESH',
                   desiredStates: ['FRESH', 'DIRTY']
               }
           }
       }
   }
})

describe('LibraryAsset context provider', () => {

    beforeEach(() => {
        store.clearActions()
        jest.clearAllMocks()
        jest.resetAllMocks()
        wmlQueryFromCacheMock.mockReturnValue(
            new WMLQuery(currentWML)
        )
    })

    it('should provide currentWML', () => {
        const TestComponent = () => {
            const { currentWML } = useLibraryAsset()
            return <div>{ currentWML }</div>
        }
        const { container } = render(
            <Provider store={store}>
                <LibraryAsset assetKey='Test' >
                    <TestComponent />
                </LibraryAsset>
            </Provider>
        )
        expect(container.textContent).toMatchSnapshot()

    })

    it('should provide normalForm', () => {
        const TestComponent = () => {
            const { normalForm } = useLibraryAsset()
            return <div>{ JSON.stringify(normalForm, null, 4) }</div>
        }
        const { container } = render(
            <Provider store={store}>
                <LibraryAsset assetKey='Test' >
                    <TestComponent />
                </LibraryAsset>
            </Provider>
        )
        expect(container.textContent).toMatchSnapshot()

    })

    it('should provide rooms', ()=>{
        const RoomComponent = ({ value }: { value: AssetComponent }) => {
            const { name, render } = value
            return <div>
                <span>{name.map((item) => ((item.tag === 'String') ? item.value : ''))}</span>
                { render.map((room, index) => {
                    switch(room.tag) {
                        case 'String':
                            return <React.Fragment key={index}>
                                <span>String</span>
                                <span>{ room.value }</span>
                            </React.Fragment>
                        case 'Link': 
                            return <React.Fragment key={index}>
                                <span>Link</span>
                                <span>{ room.to }</span>
                            </React.Fragment>
                        default:
                            return null
                    }
                })}
            </div>
        }
        const TestComponent = () => {
            const { rooms } = useLibraryAsset()
            return <div>{ Object.entries(rooms).map(([key, value]) => (<RoomComponent key={key} value={value} />)) }</div>
        }
        const { container } = render(
            <Provider store={store}>
                <LibraryAsset assetKey='Test' >
                    <TestComponent />
                </LibraryAsset>
            </Provider>
        )
        expect(container).toMatchSnapshot()

    })

    it('should provide features', ()=>{
        const FeatureComponent = ({ value }: { value: AssetComponent }) => {
            const { name, render } = value
            return <div>
                <span>{name.map((item) => ((item.tag === 'String') ? item.value : ''))}</span>
                { render.map((room, index) => {
                    switch(room.tag) {
                        case 'String':
                            return <React.Fragment key={index}>
                                <span>String</span>
                                <span>{ room.value }</span>
                            </React.Fragment>
                        case 'Link': 
                            return <React.Fragment key={index}>
                                <span>Link</span>
                                <span>{ room.to }</span>
                            </React.Fragment>
                        default:
                            return null
                    }
                })}
            </div>
        }
        const TestComponent = () => {
            const { features } = useLibraryAsset()
            return <div>{ Object.entries(features).map(([key, value]) => (<FeatureComponent key={key} value={value} />)) }</div>
        }
        const { container } = render(
            <Provider store={store}>
                <LibraryAsset assetKey='Test' >
                    <TestComponent />
                </LibraryAsset>
            </Provider>
        )
        expect(container).toMatchSnapshot()

    })

})
