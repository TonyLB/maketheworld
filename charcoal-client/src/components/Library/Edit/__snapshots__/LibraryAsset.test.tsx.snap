// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LibraryAsset context provider should provide currentWML 1`] = `
"
<Asset key=(Test) fileName=\\"test\\">
    <Import from=(BASE)>
        <Use key=(basePower) type=\\"Variable\\" as=(power) />
        <Use key=(DEF) type=\\"Room\\" />
    </Import>
    <Feature key=(clockTower)>
        <Description>
            A worn stone clock tower.
        </Description>
    </Feature>
    <Room key=(ABC)>
        <Name>Vortex</Name>
        <Description>
            Vortex
            <Link to=(toggleOpen)>(toggle)</Link>
        </Description>
    </Room>
    <If {open}>
        <Room key=(ABC)>
            <Exit to=(DEF)>welcome</Exit>
        </Room>
    </If>
    <Room key=(DEF)>
        <Name>Welcome</Name>
        <Exit to=(ABC)>vortex</Exit>
    </Room>
    <Variable key=(open) default={false} />
    <Action key=(toggleOpen) src={open = !open} />
</Asset>
"
`;

exports[`LibraryAsset context provider should provide features 1`] = `
<div>
  <div>
    <div>
      <span />
      <span>
        String
      </span>
      <span>
        A worn stone clock tower.
      </span>
    </div>
  </div>
</div>
`;

exports[`LibraryAsset context provider should provide normalForm 1`] = `
"{
    \\"Test\\": {
        \\"key\\": \\"Test\\",
        \\"tag\\": \\"Asset\\",
        \\"instance\\": false,
        \\"Story\\": false,
        \\"fileName\\": \\"test\\",
        \\"appearances\\": [
            {
                \\"contextStack\\": [],
                \\"contents\\": [
                    {
                        \\"key\\": \\"Import-0\\",
                        \\"tag\\": \\"Import\\",
                        \\"index\\": 0
                    },
                    {
                        \\"key\\": \\"clockTower\\",
                        \\"tag\\": \\"Feature\\",
                        \\"index\\": 0
                    },
                    {
                        \\"key\\": \\"ABC\\",
                        \\"tag\\": \\"Room\\",
                        \\"index\\": 0
                    },
                    {
                        \\"key\\": \\"If-0\\",
                        \\"tag\\": \\"If\\",
                        \\"index\\": 0
                    },
                    {
                        \\"key\\": \\"DEF\\",
                        \\"tag\\": \\"Room\\",
                        \\"index\\": 1
                    },
                    {
                        \\"key\\": \\"open\\",
                        \\"tag\\": \\"Variable\\",
                        \\"index\\": 0
                    },
                    {
                        \\"key\\": \\"toggleOpen\\",
                        \\"tag\\": \\"Action\\",
                        \\"index\\": 0
                    }
                ]
            }
        ]
    },
    \\"Import-0\\": {
        \\"key\\": \\"Import-0\\",
        \\"tag\\": \\"Import\\",
        \\"from\\": \\"BASE\\",
        \\"mapping\\": {
            \\"power\\": {
                \\"key\\": \\"basePower\\",
                \\"type\\": \\"Variable\\"
            },
            \\"DEF\\": {
                \\"key\\": \\"DEF\\",
                \\"type\\": \\"Room\\"
            }
        },
        \\"appearances\\": [
            {
                \\"contextStack\\": [
                    {
                        \\"key\\": \\"Test\\",
                        \\"tag\\": \\"Asset\\",
                        \\"index\\": 0
                    }
                ],
                \\"contents\\": [
                    {
                        \\"key\\": \\"power\\",
                        \\"tag\\": \\"Variable\\",
                        \\"index\\": 0
                    },
                    {
                        \\"key\\": \\"DEF\\",
                        \\"tag\\": \\"Room\\",
                        \\"index\\": 0
                    }
                ]
            }
        ]
    },
    \\"power\\": {
        \\"key\\": \\"power\\",
        \\"tag\\": \\"Variable\\",
        \\"appearances\\": [
            {
                \\"contextStack\\": [
                    {
                        \\"key\\": \\"Test\\",
                        \\"tag\\": \\"Asset\\",
                        \\"index\\": 0
                    },
                    {
                        \\"key\\": \\"Import-0\\",
                        \\"tag\\": \\"Import\\",
                        \\"index\\": 0
                    }
                ],
                \\"contents\\": []
            }
        ]
    },
    \\"DEF\\": {
        \\"key\\": \\"DEF\\",
        \\"tag\\": \\"Room\\",
        \\"appearances\\": [
            {
                \\"contextStack\\": [
                    {
                        \\"key\\": \\"Test\\",
                        \\"tag\\": \\"Asset\\",
                        \\"index\\": 0
                    },
                    {
                        \\"key\\": \\"Import-0\\",
                        \\"tag\\": \\"Import\\",
                        \\"index\\": 0
                    }
                ],
                \\"contents\\": [],
                \\"render\\": [],
                \\"name\\": []
            },
            {
                \\"contextStack\\": [
                    {
                        \\"key\\": \\"Test\\",
                        \\"tag\\": \\"Asset\\",
                        \\"index\\": 0
                    }
                ],
                \\"contents\\": [
                    {
                        \\"tag\\": \\"Exit\\",
                        \\"key\\": \\"DEF#ABC\\",
                        \\"index\\": 0
                    }
                ],
                \\"render\\": [],
                \\"name\\": [
                    {
                        \\"tag\\": \\"String\\",
                        \\"value\\": \\"Welcome\\"
                    }
                ]
            }
        ]
    },
    \\"clockTower\\": {
        \\"key\\": \\"clockTower\\",
        \\"tag\\": \\"Feature\\",
        \\"appearances\\": [
            {
                \\"contextStack\\": [
                    {
                        \\"key\\": \\"Test\\",
                        \\"tag\\": \\"Asset\\",
                        \\"index\\": 0
                    }
                ],
                \\"contents\\": [],
                \\"render\\": [
                    {
                        \\"tag\\": \\"String\\",
                        \\"value\\": \\"A worn stone clock tower.\\"
                    }
                ],
                \\"name\\": []
            }
        ]
    },
    \\"ABC\\": {
        \\"key\\": \\"ABC\\",
        \\"tag\\": \\"Room\\",
        \\"appearances\\": [
            {
                \\"contextStack\\": [
                    {
                        \\"key\\": \\"Test\\",
                        \\"tag\\": \\"Asset\\",
                        \\"index\\": 0
                    }
                ],
                \\"contents\\": [],
                \\"render\\": [
                    {
                        \\"tag\\": \\"String\\",
                        \\"value\\": \\"Vortex \\"
                    },
                    {
                        \\"tag\\": \\"Link\\",
                        \\"to\\": \\"toggleOpen\\",
                        \\"text\\": \\"(toggle)\\",
                        \\"targetTag\\": \\"Action\\"
                    }
                ],
                \\"name\\": [
                    {
                        \\"tag\\": \\"String\\",
                        \\"value\\": \\"Vortex\\"
                    }
                ]
            },
            {
                \\"contextStack\\": [
                    {
                        \\"key\\": \\"Test\\",
                        \\"tag\\": \\"Asset\\",
                        \\"index\\": 0
                    },
                    {
                        \\"key\\": \\"If-0\\",
                        \\"tag\\": \\"If\\",
                        \\"index\\": 0
                    }
                ],
                \\"contents\\": [
                    {
                        \\"tag\\": \\"Exit\\",
                        \\"key\\": \\"ABC#DEF\\",
                        \\"index\\": 0
                    }
                ],
                \\"render\\": [],
                \\"name\\": []
            }
        ]
    },
    \\"If-0\\": {
        \\"key\\": \\"If-0\\",
        \\"tag\\": \\"If\\",
        \\"conditions\\": [
            {
                \\"if\\": \\"open\\",
                \\"dependencies\\": [
                    \\"open\\"
                ]
            }
        ],
        \\"appearances\\": [
            {
                \\"contextStack\\": [
                    {
                        \\"key\\": \\"Test\\",
                        \\"tag\\": \\"Asset\\",
                        \\"index\\": 0
                    }
                ],
                \\"contents\\": [
                    {
                        \\"key\\": \\"ABC\\",
                        \\"tag\\": \\"Room\\",
                        \\"index\\": 1
                    }
                ]
            }
        ]
    },
    \\"ABC#DEF\\": {
        \\"key\\": \\"ABC#DEF\\",
        \\"tag\\": \\"Exit\\",
        \\"to\\": \\"DEF\\",
        \\"from\\": \\"ABC\\",
        \\"name\\": \\"welcome\\",
        \\"appearances\\": [
            {
                \\"contextStack\\": [
                    {
                        \\"key\\": \\"Test\\",
                        \\"tag\\": \\"Asset\\",
                        \\"index\\": 0
                    },
                    {
                        \\"key\\": \\"If-0\\",
                        \\"tag\\": \\"If\\",
                        \\"index\\": 0
                    },
                    {
                        \\"key\\": \\"ABC\\",
                        \\"tag\\": \\"Room\\",
                        \\"index\\": 1
                    }
                ],
                \\"contents\\": []
            }
        ]
    },
    \\"DEF#ABC\\": {
        \\"key\\": \\"DEF#ABC\\",
        \\"tag\\": \\"Exit\\",
        \\"to\\": \\"ABC\\",
        \\"from\\": \\"DEF\\",
        \\"name\\": \\"vortex\\",
        \\"appearances\\": [
            {
                \\"contextStack\\": [
                    {
                        \\"key\\": \\"Test\\",
                        \\"tag\\": \\"Asset\\",
                        \\"index\\": 0
                    },
                    {
                        \\"key\\": \\"DEF\\",
                        \\"tag\\": \\"Room\\",
                        \\"index\\": 1
                    }
                ],
                \\"contents\\": []
            }
        ]
    },
    \\"open\\": {
        \\"key\\": \\"open\\",
        \\"tag\\": \\"Variable\\",
        \\"default\\": \\"false\\",
        \\"appearances\\": [
            {
                \\"contextStack\\": [
                    {
                        \\"key\\": \\"Test\\",
                        \\"tag\\": \\"Asset\\",
                        \\"index\\": 0
                    }
                ],
                \\"contents\\": []
            }
        ]
    },
    \\"toggleOpen\\": {
        \\"key\\": \\"toggleOpen\\",
        \\"tag\\": \\"Action\\",
        \\"src\\": \\"open = !open\\",
        \\"appearances\\": [
            {
                \\"contextStack\\": [
                    {
                        \\"key\\": \\"Test\\",
                        \\"tag\\": \\"Asset\\",
                        \\"index\\": 0
                    }
                ],
                \\"contents\\": []
            }
        ]
    }
}"
`;

exports[`LibraryAsset context provider should provide rooms 1`] = `
<div>
  <div>
    <div>
      <span>
        
        
        
        
        
        
        
      </span>
      <span>
        String
      </span>
      <span>
        A welcome area
      </span>
    </div>
    <div>
      <span>
        
        
        
        
        
        
      </span>
      <span>
        String
      </span>
      <span>
        Vortex 
      </span>
      <span>
        Link
      </span>
      <span>
        toggleOpen
      </span>
    </div>
  </div>
</div>
`;
